File: 7 kyu/Find the next perfect square!/solution.hs
module Perfect (findNextSquare) where

findNextSquare :: Integer -> Integer
findNextSquare x | x == y = z
                 | otherwise = -1
                 where (y:z:_) = dropWhile (< x) $ (^2) <$> [0..]

_______________________________________
module Perfect (findNextSquare) where

findNextSquare :: Integer -> Integer
findNextSquare x = if isSquare x then nextSquare else -1
  where
    nextSquare = (isqrt x + 1) ^ 2

isSquare :: Integer -> Bool
isSquare x = r * r == x
  where
    r = isqrt x

isqrt :: Integer -> Integer
isqrt = floor . sqrt . fromInteger

_______________________________________
module Perfect (findNextSquare) where

findNextSquare :: Integer -> Integer
findNextSquare n | isSqr = s^2
                 | otherwise = -1
  where n' = realToFrac n
        s = floor $ sqrt n' + 1
        isSqr = floor (sqrt n')^2 == n
        
_______________________________________
module Perfect (findNextSquare) where

import Data.Bool

findNextSquare :: Integer -> Integer
findNextSquare =  ($) <$> (bool (-1) <$> head . tail <*>) . (. head) . (==)
              <*> flip dropWhile ((^2) <$> [0..]) . (>)
              
_______________________________________
module Perfect (findNextSquare) where

findNextSquare :: Integer -> Integer
findNextSquare = next . properFraction . sqrt . fromIntegral
  where
    next (i,f) = if f /= 0.0
                    then -1
                    else (i + 1)^2
                    
_______________________________________
module Perfect (findNextSquare) where

findNextSquare :: Integer -> Integer
findNextSquare x
  | isInt (sqrt (fromInteger x)) = (truncate (sqrt (fromInteger x)) + 1) ^ 2
  | otherwise = -1

isInt :: Double -> Bool
isInt = (== 0.0) . snd . properFraction 

File: 7 kyu/Find the next perfect square!/solution.cpp
#include <cmath>    
long int findNextSquare(long int sq) {
  if(sqrt(sq) != (int)sqrt(sq)){return -1;}
  return  pow(sqrt(sq) + 1,2);
}

_______________________________________
#include <cmath>

long int findNextSquare(long int sq) {
  auto sqrt = std::sqrt(sq);

  return std::modf(sqrt, &sqrt) == 0 ? std::pow(sqrt + 1, 2) : -1;
}

_______________________________________
#include <cmath>
using namespace std; 

long int findNextSquare(long int sq) 
{
 if (sqrt(sq) == floor(sqrt(sq))) return pow(sqrt(sq)+1,2);  return -1 ;
}

_______________________________________
#include <math.h>
long int findNextSquare(long int sq) {
  long int a,n,i;
  i=sqrt(sq);
  for(a = i; a <= sq; a++)
    {
        if (a*a > sq) return-1;
        if (sq == a * a)
        {
         a++;
         return a=a*a;    
        }
    }
}

_______________________________________
#include <cmath>

long int findNextSquare(long int sq) {
    // Return the next square if sq if a perfect square, -1 otherwise
    double sq_root = sqrt(sq);
      if (ceil(sq_root) != floor(sq_root))
        return -1;
    else
        return (sq_root + 1) * (sq_root + 1);
}

_______________________________________
#include <cmath>

long int findNextSquare(long int sq) {
  double base = sqrt(sq);
  if (floor(base) - base != 0)
  {
    return -1;
  }
  return round(pow(base + 1, 2));
}
